### PR TITLE
Witsml20

### LINKED ISSUE #81: Support for WITSML 2.0 perforation
Some adopters need to transfer perforation data (MDs and events).

**Describe the solution you'd like**
 Since perforation data exist in WITSML2, the intent is to allow access (Read and Write) to them in association with RESQML through wellbores.

**Describe alternatives you've considered**
- Use of RESQML wellbore frame representation with time serie properties
- extra metadata

**Additional context**
The support of WITSML 1.4.1.1 will be given up.

### PR DESCRIPTION
**Pull request by mpoudret**
What does this implement/fix? Explain your changes.
---------------------------------------------------
- Add support for WITSML 2.0 perforation.
- Move from fesapi 0.14 to fesapi 0.15

Does this close any currently open issues?
------------------------------------------
Close #81 

Any relevant logs, error output, etc?
-------------------------------------
Support for WITSML 1.4.1.1 is given up.

Where has this been tested?
---------------------------
**Operating System:** Windows 10 64bits

**Platform:** Visual Studio Community 2015 and Eclipse 2018.12 for Java example testing.

### CODE REVIEW

File: cmake/FesapiJavaExample.java [hunk 1]
```diff
@@ -24,43 +24,36 @@ Licensed to the Apache Software Foundation (ASF) under one
...
+import com.f2i.energisticsStandardsApi.${FESAPI_WITSML2_0_NS}.Well;
 
 public class FesapiJavaExample {
 	/**
 	 * Loading the FesapiCpp native library
 	 */
 	static {
 		try {
-			System.loadLibrary("${CPP_LIBRARY_NAME}.${Fesapi_VERSION}");
+			System.loadLibrary("FesapiCppUnderDev.0.15.0.0");
```
Comment (philippeVerney) [inline cmake/FesapiJavaExample.java]: I don't think it should be hard coded. The previous way looks to me better.

File: cmake/FesapiJavaExample.java [hunk 2]
```diff
@@ -363,6 +354,13 @@ private static void deserialize()
     				System.out.println("Spacing in J is not constant.");
     			}
     		}
+    		
+    		// MATHIEU
```
Comment (philippeVerney) [inline cmake/FesapiJavaExample.java]: I don't think it helps final user.

File: src/nsDefinitions.h [hunk 3]
```diff
@@ -21,4 +21,4 @@ under the License.
 #define COMMON_NS common
 #define RESQML2_NS resqml2
 #define RESQML2_0_1_NS resqml2_0_1
-#define WITSML1_4_1_1_NS witsml1_4_1_1
+#define WITSML2_0_NS witsml2_0
```
Comment (philippeVerney) [inline src/nsDefinitions.h]: I think this file should be deleted from git. It should be generated by cmake.

File: src/tools/TimeTools.h [hunk 4]
```diff
@@ -27,6 +27,16 @@ namespace timeTools
 	*/
 	std::string convertMicrosecondUnixTimestampToIso(const long long & ts);
 
+	/**
+	* Convert an Unix timestamp which is given in seconds into a date in ISO format.
+	*/
+	__declspec(dllexport) std::string convertUnixTimestampToIso(const long long & ts);
```
Comment (philippeVerney) [inline src/tools/TimeTools.h]: Do we really need to export that to final users?
I hope no.
If we really need, we should reuse export/import macro.

File: cmake/FesapiJavaExample.java (hunk 1 above)
Comment (mpoudret) [reply cmake/FesapiJavaExample.java]: Fixed.

File: cmake/FesapiJavaExample.java (hunk 2 above)
Comment (mpoudret) [reply cmake/FesapiJavaExample.java]: Fixed.

File: src/nsDefinitions.h (hunk 3 above)
Comment (mpoudret) [reply src/nsDefinitions.h]: Fixed.

File: src/tools/TimeTools.h (hunk 4 above)
Comment (mpoudret) [reply src/tools/TimeTools.h]: Fixed.

File: src/tools/TimeTools.cpp [hunk 5]
```diff
@@ -60,9 +60,13 @@ std::string timeTools::convertUnixTimestampToIso(const long long & ts)
 time_t timeTools::convertIsoToUnixTimestamp(const std::string & s)
 {
 	struct tm tm;
+#if defined(__gnu_linux__) || defined(__APPLE__)
```
Comment (philippeVerney) [inline src/tools/TimeTools.cpp]: I think it is more a check on C++11 than on the OS : https://stackoverflow.com/questions/10717502/is-there-a-preprocessor-directive-for-detecting-c11x-support

Comment (philippeVerney) [reply src/tools/TimeTools.cpp]: C++11 will be the minimum version for the coming FESAPI ETP version. We will resolve that at this time.

**Pull request merged**